src/analyses/variable-sensitivity/abstract_environment.cpp

```diff
@@ -468,3 +517,210 @@ std::vector<abstract_object_pointert> eval_operands(
 
   return operands;
 }
+
+///////////
+abstract_value_pointert as_value(const abstract_object_pointert &obj)
+{
+  auto context_value =
+    std::dynamic_pointer_cast<const context_abstract_objectt>(obj);
+
+  return context_value
+           ? as_value(context_value->unwrap_context())
+           : std::dynamic_pointer_cast<const abstract_value_objectt>(obj);
+}
+
+bool is_value(const abstract_object_pointert &obj)
+{
+  return as_value(obj) != nullptr;
+}
+
+static auto inverse_operations =
+  std::map<dstringt, dstringt>{{ID_equal, ID_notequal},
+                               {ID_notequal, ID_equal},
+                               {ID_le, ID_gt},
+                               {ID_lt, ID_ge},
+                               {ID_ge, ID_lt},
+                               {ID_gt, ID_le}};
+
+exprt invert_expr(const exprt &expr)
+{
+  auto expr_id = expr.id();
+  auto inverse_operation = inverse_operations.find(expr_id);
+  if(inverse_operation == inverse_operations.end())
+    return nil_exprt();
+
+  auto relation_expr = to_binary_relation_expr(expr);
+  auto inverse_op = inverse_operation->second;
+  return binary_relation_exprt(
+    relation_expr.lhs(), inverse_op, relation_expr.rhs());
+}
+
+exprt assume_not(
+  abstract_environmentt &env,
+  const exprt &expr,
+  const namespacet &ns)
+{
+  auto not_expr = to_not_expr(expr);
+
+  auto inverse_expression = invert_expr(not_expr.op());
+  if(inverse_expression.is_not_nil())
+    return env.do_assume(inverse_expression, ns);
+
+  auto result = env.do_assume(not_expr.op(), ns);
+  if(result.is_boolean())
+    result = result.is_true() ? exprt(false_exprt()) : true_exprt();
```

Comment:
It's a cast rather than a constructor. 
The ternary operator wants both operands to have the same type, but `result.is_true() ? false_exprt() : true_exprt()` fails because `false_exprt` and `true_exprt` are different types (even though both can be converted to the `exprt`type `result` wants. By forcing the first operand to `exprt`, the second operand decays to `exprt` automatically and it all compiles and runs. 
I've replaced with a `static_cast`.

Reply:
Ahhhhhh.... yeahhhhh... ugh; the C++ type system and when subtyping relations are taken.  That is an annoying wrinkle. Thank you.

Reply:
Using `static_cast` isn't a particularly good solution either as it is an unchecked cast. `expr_checked_cast` would not currently work either. It would need its type signature tweaking slightly and `can_cast_expr<exprt>` to be defined. I suggest just using an `if` statement instead of the ternary operator here.